File: submissions.js
var express = require('express');
var pool = require('./pool');
router = express.Router();

var types = require('pg').types;
var timestampOID = 1114;
types.setTypeParser(1114, function(stringValue) {
  return stringValue;
})

router.get('/', function(req, res){
	if (Object.keys(req.query).length == 0) {
		res.sendStatus(500);
    	return;
	}
	let text = 'SELECT * FROM submissions where'
	let values = [];
	let count = 0;
	if (req.query.id) {
		count = count + 1;
		values.push(req.query.id)
		text += ` id = \$${count}`
	}
	if (req.query.author) {
		count = count + 1;
		if (values.length != 0) {
      text += ' and';
		}
		values.push(req.query.author.toLowerCase())
		text += ` author = \$${count}`
	}
	if (req.query.title) {
		count = count + 1;
		if (values.length != 0) {
      text += ' and';
		}
		values.push(req.query.title)
		text += ` title = \$${count}`
	}
	if (req.query.subreddit) {
		count = count + 1;
		if (values.length != 0) {
      text += ' and';
		}
		values.push(req.query.subreddit.toLowerCase())
		text += ` subreddit = \$${count}`
	}
	if (req.query.after) {
		count = count + 1;
		if (values.length != 0) {
      text += ' and';
		}
		values.push(req.query.after)
		text += ` created_utc > \$${count}`
	}
	if (req.query.before) {
		count = count + 1;
		if (values.length != 0) {
      text += ' and';
		}
		values.push(req.query.before)
		text += ` created_utc < \$${count}`
	} 
	if (req.query.sort === 'ASC') {
		text += ` ORDER BY created_utc ASC`
	} else {
		text += ` ORDER BY created_utc DESC`
	}
	text += ` LIMIT 100`
	pool.query(text, values)
	.then(result => {
		res.json(result.rows);
	})
	.catch(e => {
			console.error(e.stack);
			res.sendStatus(500);
			return
	})
});

module.exports = router;

File: api/submissions.py
import json
import falcon
from psycopg2.extras import RealDictCursor

class Submissions:
   def __init__(self, pool):
      self.pool = pool

   def on_get(self, req, resp):
      text = 'SELECT * FROM submissions where'
      params = []

      if req.get_param('id'):
         text += ' id = %s'
         params.append(req.get_param('id'))

      if req.get_param('subreddit'):
         if len(params) != 0:
            text += ' and'

         params.append(str(req.get_param('subreddit')).lower())
         text += ' subreddit = %s'

      if req.get_param_as_int('after'):
         if len(params) != 0:
            text += ' and'

         params.append(req.get_param_as_int('after'))
         text += ' created_utc > %s'

      if req.get_param_as_int('before'):
         if len(params) != 0:
            text += ' and'

         params.append(req.get_param_as_int('before'))
         text += ' created_utc < %s'      

      if req.get_param('sort') == 'ASC':
         text += ' ORDER BY created_utc ASC'
      else:
         text += ' ORDER BY created_utc DESC'

      text += ' LIMIT 100'

      if len(params) == 0:
         resp.status = falcon.HTTP_500
         return

      try:
         pg_con = self.pool.getconn()
         cursor = pg_con.cursor(cursor_factory=RealDictCursor)
         cursor.execute(text, params)
         submissions = cursor.fetchall()
      except Exception as error:
         resp.status = falcon.HTTP_500
         return
      
      resp.text= json.dumps(list(submissions))
      resp.content_type = falcon.MEDIA_JSON
      resp.status = falcon.HTTP_200
